platforms/unix/src/atspi/interfaces/accessible.rs

```diff
@@ -52,62 +63,86 @@ impl AccessibleInterface<PlatformNode> {
 
     #[dbus_interface(property)]
     fn accessible_id(&self) -> ObjectId {
-        self.node.accessible_id()
+        ObjectId::from(&self.node)
     }
 
-    fn get_child_at_index(
-        &self,
-        #[zbus(header)] hdr: MessageHeader<'_>,
-        index: i32,
-    ) -> fdo::Result<(OwnedObjectAddress,)> {
+    fn get_child_at_index(&self, index: i32) -> fdo::Result<(OwnedObjectAddress,)> {
         let index = index
             .try_into()
             .map_err(|_| fdo::Error::InvalidArgs("Index can't be negative.".into()))?;
-        super::object_address(hdr.destination()?, self.node.child_at_index(index)?)
+        let child = self
+            .node
+            .child_at_index(index)
+            .map_err(self.map_error())?
+            .map(|child| ObjectId::Node {
+                adapter: self.node.adapter_id(),
+                node: child,
+            });
+        Ok(super::optional_object_address(&self.bus_name, child))
     }
 
     fn get_children(&self) -> fdo::Result<Vec<OwnedObjectAddress>> {
-        Ok(self
-            .node
-            .children()?
-            .into_iter()
-            .map(|child| child.to_address(self.bus_name.clone()))
-            .collect())
+        self.node
+            .map_children(|child| {
+                ObjectId::Node {
+                    adapter: self.node.adapter_id(),
+                    node: child,
+                }
+                .to_address(self.bus_name.clone())
+            })
+            .map_err(self.map_error())
     }
 
     fn get_index_in_parent(&self) -> fdo::Result<i32> {
-        self.node.index_in_parent()
+        self.node.index_in_parent().map_err(self.map_error())
     }
 
     fn get_role(&self) -> fdo::Result<Role> {
-        self.node.role()
+        self.node.role().map_err(self.map_error())
     }
 
     fn get_localized_role_name(&self) -> fdo::Result<String> {
-        self.node.localized_role_name()
+        self.node.localized_role_name().map_err(self.map_error())
     }
 
     fn get_state(&self) -> fdo::Result<StateSet> {
-        self.node.state()
+        self.node.state().map_err(self.map_error())
     }
 
-    fn get_application(
-        &self,
-        #[zbus(header)] hdr: MessageHeader<'_>,
-    ) -> fdo::Result<(OwnedObjectAddress,)> {
-        super::object_address(hdr.destination()?, Some(ObjectId::Root))
+    fn get_application(&self) -> (OwnedObjectAddress,) {
+        (ObjectId::Root.to_address(self.bus_name.clone()),)
     }
 
     fn get_interfaces(&self) -> fdo::Result<InterfaceSet> {
-        self.node.interfaces()
+        self.node.interfaces().map_err(self.map_error())
+    }
+}
+
+pub(crate) struct RootAccessibleInterface {
+    bus_name: OwnedUniqueName,
+    desktop_address: OwnedObjectAddress,
```

Comment:
The documentation for the `Parent` property now has this to say:

> Null parent: If the object has no parent (e.g. the application's root object is
> being queried), return "" for the application name name and "/org/a11y/atspi/null"
> for the object path.

But not all toolkits behave the same here. If we keep this desktop object around, then we'll need access to it when we'll implement the `org.a11y.atspi.Cache` interface.

Reply:
If my understanding of `atspi_accessible_get_parent` (and the functions it calls) in libatspi is correct, then it doesn't have any way of translating a null return value, as described above, into the desktop object. So either my understanding is incomplete, or that documentation is wrong. @federicomenaquintero Can you please look at this (not the whole PR, but this specific issue)?

Reply:
The documentation is clearly incorrect, no doubt about that.

So when an AT will call our cache interface, we'll have to return the desktop as the parent of the root.

Reply:
I wouldn't want to block the PR because of this. We'll move the desktop object around once we actually implement support for the cache interface.